src/org/zaproxy/zap/extension/tokengen/TokenGenerator.java

```diff
@@ -64,6 +66,9 @@ protected Void doInBackground() throws Exception {
 			try {
 				msg.getRequestHeader().setHeader(HttpHeader.COOKIE, null);
 				this.getHttpSender().sendAndReceive(msg, true);
+			} catch (SocketTimeoutException ste) {
+				log.debug("A timout occurred while sending a request to generate a token. Reducing sent count, initiating supplemental request.");
+				i--;

```

Comment:
Yeah, `MessageSummary` can check if the response is empty or not (and expose that for progress check).